File: jonghyun/11057 오르막 수.cpp
#include <iostream>
#include <vector>
#include <array>

using namespace std;

int dp(vector<array<int ,10>> &save, int n, int dec){
	if(save[n][dec]!=0){
		return save[n][dec];
	}
	
	if(n==0){
		save[n][dec]=1;
	}
	else{
		int sum=0;
		for(int i=0;i<=dec;i++){
			sum+=dp(save, n-1, i);
			sum%=10007;
		}
		save[n][dec]=sum;
	}
	
	return save[n][dec];
}

int main(){
	int n;
	cin>>n;
	
	array<int, 10> row={0};
	vector<array<int, 10>> save(n,row);
	
	int result_sum=0;
	for(int i=0;i<10;i++){
		result_sum+=dp(save, n-1, i);
	}
	
	cout<<result_sum%10007<<"\n";
	return 0;
}

File: jonghyun/11057 오르막 수.py
def dp(save, n, dec):
    if save[n][dec] != 0:
        return save[n][dec]

    if n == 0:
        save[n][dec] = 1;
    else:
        sum = 0;
        for i in range(0,dec+1):
            sum+=dp(save, n-1, i);
            sum%=10007;
        save[n][dec]=sum;
    
    return save[n][dec]; 

if __name__=="__main__":
    n=int(input())
    
    save=[[0 for j in range(10)] for i in range(n+1)]
    
    result_sum=0
    version="bottom_up"
    
    if version=="top_down":
        for i in range(10):
            result_sum+=dp(save, n-1, i)
    
    elif version=="bottom_up":
        for i in range(n):
            for dec in range(10):
                dp(save, i, dec)
        for i in range(10):
            result_sum+=dp(save, n-1, i)
    
    print(result_sum%10007)